File: mkdocs/snippets/devbook/chain/chain_heights.py
import json
import os
import time
import urllib.request

NODE_URL = os.getenv('NODE_URL', 'https://reference.symboltest.net:3001')
print(f'Using node {NODE_URL}')

prev_height = None
prev_finalized_height = None
height_changed_at = None
finalized_changed_at = None

try:
	while True:
		with urllib.request.urlopen(f'{NODE_URL}/chain/info') as response:  # [>step-1]
			chain_info = json.loads(response.read().decode())

		height = int(chain_info['height'])
		finalized = chain_info['latestFinalizedBlock']
		finalized_height = int(finalized['height'])
		# [<step-1]
		now = time.time()
		# [>step-2]
		if prev_height is not None and height != prev_height:
			height_changed_at = now
		if (
			prev_finalized_height is not None
			and finalized_height != prev_finalized_height
		):
			finalized_changed_at = now

		if height_changed_at is not None:
			height_ago = f'{int(now - height_changed_at)}s ago'
		else:
			height_ago = '-'
		if finalized_changed_at is not None:
			finalized_ago = f'{int(now - finalized_changed_at)}s ago'
		else:
			finalized_ago = '-'  # [<step-2]
		# [>step-3]
		print(
			f'Height: {height:>10,}'
			f'  (changed {height_ago})'
			f'  |  Finalized: {finalized_height:>10,}'
			f'  (changed {finalized_ago})'
		)

		prev_height = height
		prev_finalized_height = finalized_height
		time.sleep(1)
		# [<step-3]
except KeyboardInterrupt:
	pass
except Exception as error:
	print(error)

File: mkdocs/snippets/devbook/chain/ChainHeights.java
//JAVA 21+
//DEPS org.symbol:symbol-sdk:3.3.1

import java.io.IOException;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.net.http.HttpResponse.BodyHandlers;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;

public final class ChainHeights {
	private static final ObjectMapper JSON_MAPPER = new ObjectMapper();

	private static final HttpClient HTTP_CLIENT =
		HttpClient.newHttpClient();

	private final String nodeUrl = System.getenv().getOrDefault(
		"NODE_URL", "https://reference.symboltest.net:3001");

	public static void main(final String[] args) {
		try {
			new ChainHeights().run();
		} catch (final Exception ex) {
			System.out.println(null == ex.getMessage()
				? ex.toString()
				: ex.getMessage());
		}
	}

	private void run() throws IOException, InterruptedException {
		System.out.printf("Using node %s%n", nodeUrl);

		Long prevHeight = null;
		Long prevFinalizedHeight = null;
		Long heightChangedAt = null;
		Long finalizedChangedAt = null;

		while (true) {
			// [>step-1]
			final HttpRequest request = HttpRequest.newBuilder(
				URI.create(nodeUrl + "/chain/info")).GET().build();
			final HttpResponse<String> response = HTTP_CLIENT.send(
				request, BodyHandlers.ofString());
			final JsonNode chainInfo = JSON_MAPPER.readTree(
				response.body());

			final long height = chainInfo.get("height").asLong();
			final JsonNode finalized =
				chainInfo.get("latestFinalizedBlock");
			final long finalizedHeight =
				finalized.get("height").asLong();
			// [<step-1]
			final long now = System.currentTimeMillis();
			// [>step-2]
			if (null != prevHeight && height != prevHeight)
				heightChangedAt = now;
			if (null != prevFinalizedHeight
				&& finalizedHeight != prevFinalizedHeight)
				finalizedChangedAt = now;

			final String heightAgo = null != heightChangedAt
				? "%ds ago".formatted((now - heightChangedAt) / 1000)
				: "-";
			final String finalizedAgo = null != finalizedChangedAt
				? "%ds ago".formatted((now - finalizedChangedAt) / 1000)
				: "-"; // [<step-2]
			// [>step-3]
			System.out.printf(
				"Height: %,10d  (changed %s)"
					+ "  |  Finalized: %,10d  (changed %s)%n",
				height, heightAgo, finalizedHeight, finalizedAgo);

			prevHeight = height;
			prevFinalizedHeight = finalizedHeight;
			Thread.sleep(1000);
			// [<step-3]
		}
	}
}
